modules/intensity_transform/src/BIMEF.cpp

```diff
@@ -0,0 +1,562 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "precomp.hpp"
+
+#ifdef HAVE_EIGEN
+#include <Eigen/Sparse>
+#include <opencv2/core/eigen.hpp>
+#include <opencv2/imgproc.hpp>
+#endif
+
+namespace cv {
+namespace intensity_transform {
+
+#ifdef HAVE_EIGEN
+namespace
+{
+void diff(const Mat_<float>& src, Mat_<float>& srcVDiff, Mat_<float>& srcHDiff)
+{
+    srcVDiff = Mat_<float>(src.size());
+    for (int i = 0; i < src.rows; i++)
+    {
+        if (i < src.rows-1)
+        {
+            for (int j = 0; j < src.cols; j++)
+            {
+                srcVDiff(i,j) = src(i+1,j) - src(i,j);
+            }
+        }
+        else
+        {
+            for (int j = 0; j < src.cols; j++)
+            {
+                srcVDiff(i,j) = src(0,j) - src(i,j);
+            }
+        }
+    }
+
+    srcHDiff = Mat_<float>(src.size());
+    for (int j = 0; j < src.cols-1; j++)
+    {
+        for (int i = 0; i < src.rows; i++)
+        {
+            srcHDiff(i,j) = src(i,j+1) - src(i,j);
+        }
+    }
+    for (int i = 0; i < src.rows; i++)
+    {
+        srcHDiff(i,src.cols-1) = src(i,0) - src(i,src.cols-1);
+    }
+}
+
+void computeTextureWeights(const Mat_<float>& x, float sigma, float sharpness, Mat_<float>& W_h, Mat_<float>& W_v)
+{
+    Mat_<float> dt0_v, dt0_h;
+    diff(x, dt0_v, dt0_h);
+
+    Mat_<float> gauker_h;
+    Mat_<float> kernel_h = Mat_<float>::ones(1, static_cast<int>(sigma));
+    filter2D(dt0_h, gauker_h, -1, kernel_h, Point(-1,-1), 0, BORDER_CONSTANT);
+
+    Mat_<float> gauker_v;
+    Mat_<float> kernel_v = Mat_<float>::ones(static_cast<int>(sigma), 1);
+    filter2D(dt0_v, gauker_v, -1, kernel_v, Point(-1,-1), 0, BORDER_CONSTANT);
+
+    W_h = Mat_<float>(gauker_h.size());
+    W_v = Mat_<float>(gauker_v.size());
+
+    for (int i = 0; i < gauker_h.rows; i++)
+    {
+        for (int j = 0; j < gauker_h.cols; j++)
+        {
+            W_h(i,j) = 1 / (std::abs(gauker_h(i,j)) * std::abs(dt0_h(i,j)) + sharpness);
+            W_v(i,j) = 1 / (std::abs(gauker_v(i,j)) * std::abs(dt0_v(i,j)) + sharpness);
+        }
+    }
+}
+
+template <class numeric_t>
+Eigen::SparseMatrix<numeric_t> spdiags(const Eigen::Matrix<numeric_t,-1,-1> &B,
```

Comment:
`spdiags` function is taken from [here](https://gitlab.math.ethz.ch/NumCSE/NumCSE/blob/9948847615504057f69eb939297121ab70149af5/LectureCodes/MatVec/Sparse/sparseformat/Eigen/spdiags.hpp).